### DISCUSSION TITLE
NGINX PHP Configuration Missing fastcgi_split_path_info

### DISCUSSION THREAD

**Discussion by renpytom** [wavelog/wavelog#1133] [General]
**This is for reporting bugs with the code, if this issue is regarding setting up on your local server please use the discussions area**

**Describe the bug**
Provide a clear and concise description of the bug.

**Version Information**
We need some information about your versions. You can get all Information in the Admin > Debug Menu
- OS:               Ubuntu 22.04
- PHP:              8.3
- MySQL/MariaDB:    10.11
- Wavelog:          1.88

**Steps to Reproduce**
Install wavelog using the instructions at:

https://github.com/wavelog/Wavelog/wiki/Installation

with the NGINX configuration at: 

https://github.com/wavelog/wavelog/wiki/Webserver-Configurations#nginx

The install process fails during the migrate step.

**Expected Behavior**

The install process succeeds. 

**Additional Context**

This seems to be because the default NGINX example doesn't split the path info from the rest of the URL. I managed to get things working by changing the location .php section to be:

```
    location ~ \.php(/|$) {
        fastcgi_pass unix:/var/run/php/php-fpm.sock;

        # 404
        try_files                     $fastcgi_script_name =404;

        # default fastcgi_params
        include                       fastcgi_params;

        # fastcgi settings
        fastcgi_index                 index.php;
        fastcgi_buffers               8 16k;
        fastcgi_buffer_size           32k;

        # fastcgi params
        fastcgi_split_path_info ^(.+\.php)(/.*)$;
        fastcgi_param DOCUMENT_ROOT   $realpath_root;
        fastcgi_param SCRIPT_FILENAME $realpath_root$fastcgi_script_name;
        fastcgi_param PATH_INFO $fastcgi_path_info;
    } 
``` 

The fastcgi_split_path_info and fastcgi_param PATH_INFO lines seem to be the most important ones here. I also think that the new location, which matches .php at the  end of a path segment, is probably better than the one that's being pointed to now. 

73, AA2TR

**Comment by HB9HIL** [comment]
Hello @renpytom 

thx for your description. 

> The install process fails during the migrate step.

This is indeed something which happens from time to time. I did not fully understand when and why this happens. Sometimes this happens on docker installs aswell, which use Apache2 per default.

> This seems to be because the default NGINX example doesn't split the path info from the rest of the URL. 

nginx is a bit more tricky here. We use the (old) Codeigniter 3 Framework. There are plenty of descriptions and reports in the www about issue with setting up nginx with Codeigniter 3. I did not found an "universal" working configuration.

> I also think that the new location, which matches .php at the end of a path segment, is probably better than the one that's being pointed to now.

I'm not sure about this. Did you change `$config['index_page']` in your `application/config/config.php` and set it to `''`?
If I want to remove index.php from the URL there won't be an `.php` at the end. Also I'm working on a new map feature where the URL does return a `.png` image (but this is not part of the URL)
You may need to test your configuration with set `$config['index_page'] = ''` in your config.php

> [!IMPORTANT]
> Please Feedback!

Because of the nature of this problem and this is no new bug I move this one to the discussions. 

If there is a bug in the installer we can link it to the existing issue: https://github.com/wavelog/wavelog/issues/1011

**Reply by HB9HIL** [reply]
PS: Regarding the Wiki. If you are **sure** that this works with different kind of configurations mentioned above, feel free to change the wiki yourself 👍  😃

**Reply by renpytom** [reply]
I was able to get things working, first with ``$config['index_page'] = 'index.php'``, and then changed it to ``$config['index_page'] = ''``.  My wavelog instance is now deployed at https://wavelog.aa2tr.org/visitor/w2lrc . 

I've also gotten several other PHP sites working with a similar configuration, so I have some confidence in it. 

I'll go ahead and edit the wiki to match my setup, but there are several places (IIRC, one during setup) that point you to make changes to your NGINX configuration that are likely to no longer be necessary.

**Reply by HB9HIL** [reply]
So you say basically `$config['index_page'] = ''` should be default if using nginx? 

@violarulan is using nginx aswell, but not sure. Also @hellofinch 
Can you too confirm that?

**Comment by violarulan** [comment]
@HB9HIL 

I have been using Nginx all the time. Just did another fresh installation test and it works. 

My full config is as follows:

```
# nginx.conf

worker_processes auto;
worker_cpu_affinity auto;

events {
    multi_accept on;
    worker_connections 1024;
}

http {
    charset utf-8;
    sendfile on;
    tcp_nopush on;
    tcp_nodelay on;
    server_tokens off;
    log_not_found off;
    types_hash_max_size 4096;
    client_max_body_size 16M;

    ssl_session_cache   shared:SSL:10m;
    ssl_session_timeout 10m;
    # MIME
    include mime.types;
    default_type application/octet-stream;

    error_log /var/log/nginx/error.log warn;

    # load configs
    include /etc/nginx/conf.d/*.conf;
    include /etc/nginx/sites-enabled/*;

    keepalive_timeout  70;

    server {
        listen       80 default_server;
        server_name  _;

        location / {
            root   /usr/share/nginx/html;
            index  index.html index.htm;
        }

        # redirect server error pages to the static page /50x.html
        #
        error_page   500 502 503 504  /50x.html;
        location = /50x.html {
            root   /usr/share/nginx/html;
        }
    }

}
```

```
# sites-enabled/logging.conf
server {
    listen 80;
    listen [::]:80;
    server_name <reduntant>;

    # Discourage deep links by using a permanent redirect to home page of HTTPS site
    return 301 https://$host;
}

server {
    listen 443 ssl;
    listen [::]:443 ssl;
    
    server_name <reduntant>;
    ssl_certificate <reduntant>;
    ssl_certificate_key <reduntant>;
    ssl_protocols       TLSv1 TLSv1.1 TLSv1.2 TLSv1.3;
    ssl_ciphers         HIGH:!aNULL:!MD5;
    # http2        on;

    add_header Strict-Transport-Security "max-age=31536000; includeSubDomains" always;
    root /<reduntant>;

    location ~* \.(?:css(\.map)?|js(\.map)?|jpe?g|png|gif|ico|cur|heic|woff|woff2|webp|tiff?|mp3|m4a|aac|ogg|midi?|wav|mp4|mov|webm|mpe?g|avi|ogv|flv|wmv)$ {
        try_files $uri $uri/;
    }
    
    location /.git {
        return 404;
    }

    location / { 
        try_files $uri $uri/ /index.php?$args;        
        # default fastcgi_params
        include fastcgi_params;

        # fastcgi settings
        fastcgi_pass			unix:/run/php-fpm/php-fpm.sock;
        fastcgi_index			index.php;
        fastcgi_buffers			8 16k;
        fastcgi_buffer_size		32k;

        # fastcgi params
        fastcgi_param DOCUMENT_ROOT	$realpath_root;
        fastcgi_param SCRIPT_FILENAME	$realpath_root$fastcgi_script_name;
        #fastcgi_param PHP_ADMIN_VALUE	"open_basedir=$base/:/usr/lib/php/:/tmp/";
    }
}
```

```
# fastcgi_params

fastcgi_param  QUERY_STRING       $query_string;
fastcgi_param  REQUEST_METHOD     $request_method;
fastcgi_param  CONTENT_TYPE       $content_type;
fastcgi_param  CONTENT_LENGTH     $content_length;

fastcgi_param  SCRIPT_NAME        $fastcgi_script_name;
fastcgi_param  REQUEST_URI        $request_uri;
fastcgi_param  DOCUMENT_URI       $document_uri;
fastcgi_param  DOCUMENT_ROOT      $document_root;
fastcgi_param  SERVER_PROTOCOL    $server_protocol;
fastcgi_param  REQUEST_SCHEME     $scheme;
fastcgi_param  HTTPS              $https if_not_empty;

fastcgi_param  GATEWAY_INTERFACE  CGI/1.1;
fastcgi_param  SERVER_SOFTWARE    nginx/$nginx_version;

fastcgi_param  REMOTE_ADDR        $remote_addr;
fastcgi_param  REMOTE_PORT        $remote_port;
fastcgi_param  SERVER_ADDR        $server_addr;
fastcgi_param  SERVER_PORT        $server_port;
fastcgi_param  SERVER_NAME        $server_name;

# PHP only, required if PHP was built with --enable-force-cgi-redirect
fastcgi_param  REDIRECT_STATUS    200;
```

In my setup, I make `/` handle everything except the static assets. It works under both condition (Pretty and normal URL). 

If I would like to use pretty URL, I just make it works by changing the `$config['index_page'] = 'index.php'` to `''`, without changing nginx config.

**Comment by hellofinch** [comment]
@HB9HIL 
Hello, I'm not an expert on nginx. But I did another fresh installation test following the installation guide and it works.
I deploy the wavelog based on docker image [fbraz3/lnmp:8.3](https://hub.docker.com/r/fbraz3/lnmp) which frees me from installing and config nginx and database.
First, without any modifications in the nginx config. I finished the installation but failed to redirect to the login page. (This is normal for me and I don't know why. The database and everything look fine.)
Then, I should modify the nginx configuration. `try_files $uri $uri/ /wavelog/index.php?$query_string;`. This is one of important line for me. `$query_string` is necessary. This argument came from CodeIgniter. Because I read the [guide](https://www.codeigniter.com/user_guide/general/urls.html#nginx) of CodeIgniter4, there is a part to explain why the argument should be there. I think the guide is helpful.
Next, I manually visited the login page with `IP:PORT/wavelog/`. It will fail. So I should modify the `application/config/config.php`. I change `$config['index_page'] = 'index.php';` to `$config['index_page'] = '';`. If I don't do so, I will stuck in a redirect loop as I mentioned before.
Finally, I see the dashboard page. But there still are some errors. The page continues to append the same info at the end of the page. The web will stuck in a few minutes. So I should modify the Ngnix configuration again.
I change `location ~ \.php {` to `location ~ \.php$ {`. The `$` is important for me. After modifying, restart the Nginx. Everything works fine.
Here is my configuration of Nginx.
```
server {
    listen 80 default_server;

    client_max_body_size 100M;

    root /app/public;
    index index.php index.html index.htm;

    location /pma/ {
        root /var/www/html;
        index index.php;

        location ~ \.php$ {
            fastcgi_buffer_size 128k;
            fastcgi_buffers 4 256k;
            fastcgi_busy_buffers_size 256k;

            fastcgi_pass unix:/run/php/php8.3-fpm.sock;
            fastcgi_param SCRIPT_FILENAME $document_root$fastcgi_script_name;
            fastcgi_param SCRIPT_NAME $fastcgi_script_name;
            include /etc/nginx/fastcgi_params;
            fastcgi_read_timeout 300;
        }
    }
    location /wavelog/ {
        root /var/www/html;
        try_files $uri $uri/ /wavelog/index.php?$query_string; #$query_string is important for me.
        location ~ \.php$ { #$ is important for me.
            fastcgi_index index.php;
            fastcgi_buffer_size 128k;
            fastcgi_buffers 4 256k;
            fastcgi_busy_buffers_size 256k;

            fastcgi_pass unix:/run/php/php8.3-fpm.sock;
            fastcgi_param DOCUMENT_ROOT   $realpath_root;
            fastcgi_param SCRIPT_FILENAME $document_root$fastcgi_script_name;
            fastcgi_param SCRIPT_NAME $realpath_root$fastcgi_script_name;
            include /etc/nginx/fastcgi_params;
            fastcgi_read_timeout 3600;
        }
    }
    location / {
        try_files $uri $uri/ /index.php$is_args$args;
    }

    location ~ \.php$ { 
        fastcgi_buffer_size 128k;
        fastcgi_buffers 4 256k;
        fastcgi_busy_buffers_size 256k;

        fastcgi_pass unix:/run/php/php8.3-fpm.sock;
        fastcgi_param SCRIPT_FILENAME $document_root$fastcgi_script_name;
        fastcgi_param SCRIPT_NAME $fastcgi_script_name;
        include /etc/nginx/fastcgi_params;
        fastcgi_read_timeout 300;
    }

    location ~ /\.ht {
        deny all;
    }
}
```
the wavelog code is located in `/var/www/html/wavelog`.
:) 73

**Reply by hellofinch** [reply]
BTW. If necessary, I can try to export the docker container. Maybe it can help you to figure it out.
:)
